controllers/machineset_controller_test.go

```diff
@@ -181,6 +182,182 @@ var _ = Describe("MachineSet Reconciler", func() {
 	})
 })
 
+func TestMachineSetOwnerReference(t *testing.T) {
+	ml := &clusterv1.MachineList{}
+
+	clusterIncorrectMeta := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       "my-kind",
+			"apiVersion": "my-api-version",
+			"metadata": map[string]interface{}{
+				"name":      "invalid-cluster",
+				"namespace": "default",
+			},
+		},
+	}
+
+	clusterCorrectMeta := &clusterv1.Cluster{
+		TypeMeta:   metav1.TypeMeta{Kind: "Cluster", APIVersion: clusterv1.GroupVersion.String()},
+		ObjectMeta: metav1.ObjectMeta{Namespace: "default", Name: "valid-cluster"},
+	}
+
+	ms1 := newMachineSet("machineset1", "valid-cluster")
+	ms2 := newMachineSet("machineset2", "invalid-cluster")
+	ms3 := newMachineSet("machineset3", "valid-cluster")
+	ms3.OwnerReferences = []metav1.OwnerReference{
+		{
+			APIVersion: clusterv1.GroupVersion.String(),
+			Kind:       "MachineDeployment",
+			Name:       "valid-machinedeployment",
+		},
+	}
+
+	testCases := []struct {
+		name               string
+		request            reconcile.Request
+		ms                 *clusterv1.MachineSet
+		expectReconcileErr bool
+		expectedOR         []metav1.OwnerReference
+	}{
+		{
+			name: "should add cluster owner reference to machine set",
+			request: reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      ms1.Name,
+					Namespace: ms1.Namespace,
+				},
+			},
+			ms: ms1,
+			expectedOR: []metav1.OwnerReference{
+				{
+					APIVersion: clusterCorrectMeta.APIVersion,
+					Kind:       clusterCorrectMeta.Kind,
+					Name:       clusterCorrectMeta.Name,
+					UID:        clusterCorrectMeta.UID,
+				},
+			},
+		},
+		{
+			name: "should add not add cluster owner reference with incorrect meta",
+			request: reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      ms2.Name,
+					Namespace: ms2.Namespace,
+				},
+			},
+			ms:                 ms2,
+			expectReconcileErr: true,
+		},
+		{
+			name: "should not add cluster owner reference if machine is owned by a machine deployment",
+			request: reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      ms3.Name,
+					Namespace: ms3.Namespace,
+				},
+			},
+			ms: ms3,
+			expectedOR: []metav1.OwnerReference{
+				{
+					APIVersion: clusterv1.GroupVersion.String(),
+					Kind:       "MachineDeployment",
+					Name:       "valid-machinedeployment",
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			RegisterTestingT(t)
```

Comment:
I added the `RegisterTestingT` here instead of the top because this `t` that asserts failure is different from the one above.
If there was only a single `RegisterTestingT` used above we would get the following error on test failure:
```
test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
```
This is further explained here: https://github.com/golang/mock/issues/51#issuecomment-324423178